### PR TITLE
add with_data and sample uses

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
state(), set_state() (and serialize/deserialize_data) end up copy-pasta across code trying to more natively interact with account objects

#### Summary of Changes
add with_data(), which takes a Result closure
added some example use cases

feedback appreciated on errors.  I tried a bunch of different ways to show plumb bincode::ErrorKind errors and E errors, and this seemed like the least painful.  Box<std::error::Error> was much prettier, but because Solana doesn't use Boxed errors, the compatibility broke right at process_instruction()

### CODE REVIEW

**Comment by rob-solana** [comment]
@garious @sagar-solana

File: sdk/src/account_utils.rs [hunk 1]
```diff
@@ -1,11 +1,13 @@
 //! useful extras for Account state
 use crate::account::{Account, KeyedAccount};
 use crate::instruction::InstructionError;
-use bincode::ErrorKind;
 
 /// Convenience trait to covert bincode errors to instruction errors.
 pub trait State<T> {
     fn state(&self) -> Result<T, InstructionError>;
+    //    fn with_state<R, E, F>(&mut self, func: F) -> Result<Result<R, E>, InstructionError>
```
Comment (sagar-solana) [inline sdk/src/account_utils.rs]: unused?

File: sdk/src/instruction.rs [hunk 2]
```diff
@@ -58,6 +58,27 @@ pub enum InstructionError {
     CustomError(u32),
 }
 
+impl std::error::Error for InstructionError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        None
+    }
+}
+
+impl std::fmt::Display for InstructionError {
```
Comment (sagar-solana) [inline sdk/src/instruction.rs]: Can we just #derive Debug for InstructionError?

**Comment by sagar-solana** [comment]
Yeah I think I like it. Maybe `with_data` should be called `update` or something.

File: sdk/src/instruction.rs (hunk 2 above)
Comment (rob-solana) [reply sdk/src/instruction.rs]: we want to do that, too.  having Display is a requirement of std::error::Error, apparently

File: sdk/src/account_utils.rs (hunk 1 above)
Comment (rob-solana) [reply sdk/src/account_utils.rs]: I am holding off putting these in, because with account.rs support, this file actually goes away...

**Comment by rob-solana** [comment]
from my digging around in stuff online, looks like the "right" way to deal with errors propagating through layers is to `Box` 'em up.  this is a longer-pole job for us, we'd maybe need to touch a lot of `Result<>` code (though we might not)

**Comment by rob-solana** [comment]
closing until we're boxing errors from instruction_processor()

**Pull request closed without merging**